binrw_derive/src/codegen/typed_builder.rs

```diff
@@ -0,0 +1,248 @@
+use proc_macro2::TokenStream;
+use quote::{quote, ToTokens};
+use syn::{Ident, Type};
+
+#[allow(clippy::wildcard_imports)]
+use crate::codegen::sanitization::*;
+use crate::parser::meta_types::NamedImport;
+
+pub(crate) enum BuilderFieldKind {
+    Required,
+    Optional { default: Box<syn::Expr> },
+}
+
+pub(crate) struct BuilderField {
+    pub(crate) name: Ident,
+    pub(crate) ty: Type,
+    pub(crate) kind: BuilderFieldKind,
+}
+
+pub(crate) struct Builder<'a> {
+    pub(crate) builder_name: &'a Ident,
+    pub(crate) result_name: &'a Ident,
+    pub(crate) fields: &'a [BuilderField],
+}
+
+impl<'a> Builder<'a> {
+    pub(crate) fn generate(&self) -> TokenStream {
+        let builder_name = self.builder_name;
+        let name = self.result_name;
+        let fields = self.generate_result_fields();
+        let builder_fields = self.generate_builder_fields();
+        let initial = self.generate_builder_initial();
+        let generics = self.generate_generics();
+        let initial_generics = self.generate_initial_generics();
+        let setters = self.generate_setters();
+        let satisfied = std::iter::repeat(SATISFIED_OR_OPTIONAL);
```

Comment:
It is used inside a repetition block, so it needs to be an iterator

Reply:
I don’t think that’s correct: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=9162647f7a6085068b5c4bad41b76cec

Reply:
I've had mixed results when it comes to this. Sometimes it works immediately, sometimes I need to create a local variable and then it works. I don't remember needing to create an iterator, but it's been awhile since i touched the proc macro code.

Reply:
> I don’t think that’s correct: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=9162647f7a6085068b5c4bad41b76cec

Hrm. I guess it's because it is a constant? I'm not entirely sure why I was getting an error on it.